config.org
#+title: Config
#+PROPERTY: header-args :tangle config.el

现在我用的主题是……

#+begin_src emacs-lisp
(setq doom-theme 'ef-arbutus)
#+end_src

* 启动
今日方知我是我：

#+begin_src emacs-lisp
(setq user-full-name "SouthFox"
      user-mail-address "master@southfox.me")
#+end_src

** 横幅
替换 =Doom Emacs= 的横幅，用一只狐狸！

#+begin_src emacs-lisp
(defun doom-dashboard-draw-ascii-emacs-banner-fn ()
  (let* ((banner
          '("                                                                   ,           "
            "                                                             _.-=;~ /_         "
            "                                                          _-~   '     ;.       "
            "                                                      _.-~     '   .-~-~`-._   "
            "                                                _.--~~:.             --.____88 "
            "                              ____.........--~~~. .' .  .        _..-------~~  "
            "                     _..--~~~~               .' .'             ,'              "
            "                 _.-~                        .       .     ` ,'                "
            "               .'                                    :.    ./                  "
            "             .:     ,/          `                   ::.   ,'                   "
            "           .:'     ,(            ;.                ::. ,-'                     "
            "          .'     ./'.`.     . . /:::._______.... _/:.o/                        "
            "         /     ./'. . .)  . _.,'               `88;?88|                        "
            "       ,'  . .,/'._,-~ /_.o8P'                  88P ?8b                        "
            "    _,'' . .,/',-~    d888P'                    88'  88|                       "
            " _.'~  . .,:oP'        ?88b              _..--- 88.--'8b.--..__                "
            ":     ...' 88o __,------.88o ...__..._.=~- .    `~~   `~~      ~-._ Fox! _.    "
            "`.;;;:='    ~~            ~~~                ~-    -       -   -               "))
         (longest-line (apply #'max (mapcar #'length banner))))
    (put-text-property
     (point)
     (dolist (line banner (point))
       (insert (+doom-dashboard--center
                +doom-dashboard--width
                (concat
                 line (make-string (max 0 (- longest-line (length line)))
                                   32)))
               "\n"))
     'face 'doom-dashboard-banner)))

;(unless (display-graphic-p) ; for some reason this messes up the graphical splash screen atm
  (setq +doom-dashboard-ascii-banner-fn #'doom-dashboard-draw-ascii-emacs-banner-fn) ;)
#+end_src

* org-mode
将 =org-log-into-drawer= 设为 =t= 后，完成任务后的时间戳将会收起，
这样一些每天完成的任务就不会被相关时间戳占满。

#+begin_src emacs-lisp
(after! org
  (setq org-directory "~/org/"
        org-image-actual-width '(400)
        org-link-search-must-match-exact-headline nil
        org-log-done 'time
        org-log-into-drawer t))

#+end_src

doom emacs 里的远程图片修复：

#+begin_src elisp
;; DOOM hack fix https://github.com/doomemacs/doomemacs/issues/8164
(after! org
  (defun org-image-update-overlay (file link &optional data-p refresh)
    "Create image overlay for FILE associtated with org-element LINK.
If DATA-P is non-nil FILE is not a file name but a string with the image data.
If REFRESH is non-nil don't download the file but refresh the image.
See also `create-image'.
This function is almost a duplicate of a part of `org-display-inline-images'."
    (when (or data-p (file-exists-p file))
      (let ((width
             ;; Apply `org-image-actual-width' specifications.
             (cond
              ((eq org-image-actual-width t) nil)
              ((listp org-image-actual-width)
               (or
                ;; First try to find a width among
                ;; attributes associated to the paragraph
                ;; containing link.
                (let ((paragraph
                       (let ((e link))
                         (while (and (setq e (org-element-property
                                              :parent e))
                                     (not (eq (org-element-type e)
                                              'paragraph))))
                         e)))
                  (when paragraph
                    (save-excursion
                      (goto-char (org-element-property :begin paragraph))
                      (when
                          (re-search-forward
                           "^[ \t]*#\\+attr_.*?: +.*?:width +\\(\\S-+\\)"
                           (org-element-property
                            :post-affiliated paragraph)
                           t)
                        (string-to-number (match-string 1))))))
                ;; Otherwise, fall-back to provided number.
                (car org-image-actual-width)))
              ((numberp org-image-actual-width)
               org-image-actual-width)))
            (old (get-char-property-and-overlay
                  (org-element-property :begin link)
                  'org-image-overlay)))
        (if (and (car-safe old) refresh)
            (image-flush (overlay-get (cdr old) 'display))
          (let ((image (create-image file
                                     (and (image-type-available-p 'imagemagick)
                                          width
                                          'imagemagick)
                                     data-p
                                     :width width)))
            (when image
              (let* ((link
                      ;; If inline image is the description
                      ;; of another link, be sure to
                      ;; consider the latter as the one to
                      ;; apply the overlay on.
                      (let ((parent
                             (org-element-property :parent link)))
                        (if (eq (org-element-type parent) 'link)
                            parent
                          link)))
                     (ov (make-overlay
                          (org-element-property :begin link)
                          (progn
                            (goto-char
                             (org-element-property :end link))
                            (skip-chars-backward " \t")
                            (point)))))
                (overlay-put ov 'display image)
                (overlay-put ov 'face 'default)
                (overlay-put ov 'org-image-overlay t)
                (overlay-put
                 ov 'modification-hooks
                 (list 'org-display-inline-remove-overlay))
                (push ov org-inline-image-overlays)
                ov)))))))

  (defun org-display-user-inline-images (&optional _include-linked _refresh beg end)
    "Like `org-display-inline-images' but for image data links.
_INCLUDE-LINKED and _REFRESH are ignored.
Restrict to region between BEG and END if both are non-nil.
Image data links have a :image-data-fun parameter.
\(See `org-link-set-parameters'.)
The value of the :image-data-fun parameter is a function
taking the PROTOCOL, the LINK, and the DESCRIPTION as arguments.
If that function returns nil the link is not interpreted as image.
Otherwise the return value is the image data string to be displayed.

Note that only bracket links are allowed as image data links
with one of the formats
 [[PROTOCOL:LINK]]
or
 [[PROTOCOL:LINK][DESCRIPTION]]
are recognized."
    (interactive)
    (when (and (called-interactively-p 'any)
               (use-region-p))
      (setq beg (region-beginning)
            end (region-end)))
    (when (display-graphic-p)
      (org-with-wide-buffer
       (goto-char (or beg (point-min)))
       (when-let ((image-data-link-parameters
		   (cl-loop for link-par-entry in org-link-parameters
			    with fun
			    when (setq fun (plist-get (cdr link-par-entry) :image-data-fun))
			    collect (cons (car link-par-entry) fun)))
		  (image-data-link-re (regexp-opt (mapcar 'car image-data-link-parameters)))
		  (re (format "\\[\\[\\(%s\\):\\([^]]+\\)\\]\\(?:\\[\\([^]]+\\)\\]\\)?\\]"
			      image-data-link-re)))
         (while (re-search-forward re end t)
           (let* ((protocol (match-string-no-properties 1))
		  (link (match-string-no-properties 2))
		  (description (match-string-no-properties 3))
		  (image-data-link (assoc-string protocol image-data-link-parameters))
		  (el (save-excursion (goto-char (match-beginning 1)) (org-element-context)))
		  image-data)
             (when el
               (setq image-data
                     (or (let ((old (get-char-property-and-overlay
                                     (org-element-property :begin el)
                                     'org-image-overlay)))
                           (and old
                                (car-safe old)
                                (overlay-get (cdr old) 'display)))
		         (funcall (cdr image-data-link) protocol link description)))
               (when image-data
                 (let ((ol (org-image-update-overlay image-data el t t)))
                   (when (and ol description)
                     (overlay-put ol 'after-string description)))))))))))

  (advice-add #'org-display-inline-images :after #'org-display-user-inline-images))
#+end_src

** org-capture
一些捕获模板和路径设置，但说实在自从用了 =org-roam= 那边这边就没怎么管了。

#+begin_src emacs-lisp
(setq org-agenda-files '("~/Sync/org/GTD/inbox.org"
                         "~/Sync/org/GTD/tickler.org"))

(after! org
  (setq org-capture-templates '(("t" "Todo [inbox]" entry
                               (file "~/Sync/org/GTD/inbox.org")
                               "* TODO %i%? \n SCHEDULED: %t")
                              ("T" "Tickler" entry
                               (file+headline "~/Sync/org/GTD/tickler.org" "Tickler")
                               "* %i%? \n %U"))))
#+end_src

** org-roam
主要是一些捕获模板设置，文件命名加上时间还是因为导出到数字花园时会放在一个文件夹里，
而为了避免重名。

#+begin_src emacs-lisp
(setq org-roam-directory "~/Sync/org/Note/org-roam")
(setq org-roam-capture-templates
      '(("m" "main" plain
	 "%?"
	 :if-new (file+head "main/%<%Y%m%d%H%M%S>-${slug}.org"
                            "#+title: ${title}\n#+date: %T\n")
	 :immediate-finish t
	 :unnarrowed t)
	("r" "references" plain "%?"
	 :if-new
	 (file+head "references/%<%Y%m%d%H%M%S>-${title}.org" "#+title: ${title}\n#+date: %T\n")
	 :immediate-finish t
	 :unnarrowed t)
	("a" "article" plain "%?"
	 :if-new
	 (file+head "articles/%<%Y%m%d%H%M%S>-${title}.org" "#+title: ${title}\n#+date: %T\n")
	 :immediate-finish t
	 :unnarrowed t)))

(setq org-roam-dailies-capture-templates
      '(("d" "default" plain
         "* %?"
         :target (file+head "%<%Y-%m-%d>.org"
                            "#+title: %<%Y-%m-%d>\n#+date: %T\n"))))

(with-eval-after-load 'ox-hugo
  (setq org-hugo-anchor-functions '(org-hugo-get-id
                                    org-hugo-get-heading-slug
                                    org-hugo-get-md5)))
#+end_src

*** Garden
数字花园，开始种植思想……

#+begin_src emacs-lisp
(defcustom org-roam-garden-file "~/Sync/org/GTD/garden.org"
  "The file used by roam garden."
  :group 'org-roam
  :type 'string)

(add-to-list 'org-agenda-files org-roam-garden-file)
#+end_src

#+begin_src emacs-lisp
(defun org-roam-garden-plant ()
  (interactive)
  (save-excursion
    (goto-char (point-min))
    (let* ((node (org-roam-node-at-point))
           (title (org-roam-node-title node))
           (file (org-roam-node-file node)))
      (org-set-property "GARDEN-STATE" "seedling")
      (org-roam-update-garden-file title file))
    (save-buffer)))

(defun org-roam-update-garden-file (heading file-path)
  (with-current-buffer (find-file-noselect org-roam-garden-file)
    (let ((buffer (current-buffer)))
      (if (org-find-exact-headline-in-buffer heading)
          (message "Plant exists, skip...")
        (goto-char (point-max))
        (insert (concat "* " heading "\n[[" file-path "]]"))
        (org-schedule nil (format "<%s 22:00 ++3d >" (format-time-string "%Y-%m-%d %u")))
        (save-buffer)))))

(defun my/org-roam-filter-by-properties (properties-name)
  (lambda (node)
    (assoc properties-name (org-roam-node-properties node))))

(defun my/org-roam-list-notes-by-properties (properties-name)
  (seq-filter
   (my/org-roam-filter-by-properties properties-name)
   (org-roam-node-list)))

(defun my/org-roam-garden-dashboard--build-item (node)
  (concat "\n** " (org-roam-node-title node) "\n[[" (org-roam-node-file node) "]]"))

(defun my/org-roam-garden-dashboard ()
  (interactive)
  (with-current-buffer (get-buffer-create "*Garden*")
    (erase-buffer)
    (let ((filtered-node (my/org-roam-list-notes-by-properties "GARDEN-STATE")))
      (insert "* 🌱 seedling")
      (insert (string-join (mapcar #'my/org-roam-garden-dashboard--build-item
                                   filtered-node))))
    (org-mode)
    (switch-to-buffer (current-buffer))))
#+end_src

** org-fragtog
好像是 =latex= 相关的预览包……？

#+begin_src emacs-lisp
(use-package! org-fragtog
  :hook (org-mode . org-fragtog-mode))
#+end_src

** roam-publish
一些从世界各地拼出来的代码，还有一些自己单纯堆叠调用的函数，
我真该系统学学 =elisp= +1 。

#+begin_src emacs-lisp
(setq org-hugo-base-dir "~/Documents/roam-publish/")

(defun my/org-roam-filter-by-tag (tag-name)
  (lambda (node)
    (member tag-name (org-roam-node-tags node))))

(defun my/org-roam-list-notes-by-tag (tag-name)
  (mapcar #'org-roam-node-file
          (seq-filter
           (my/org-roam-filter-by-tag tag-name)
           (org-roam-node-list))))

(defun my/org-roam-export-all ()
  "Re-exports all Org-roam files to Hugo markdown."
  (interactive)
  (dolist (org-file (my/org-roam-list-notes-by-tag "publish"))
  ;(dolist (org-file (directory-files-recursively org-roam-directory "\.org$"))
    (with-current-buffer (find-file org-file)
        (my/org-roam-publish t))))

(defun my/insert--backlinks (&optional heading-func)
  (if-let* ((file-title (org-get-title))
            (backlinks (org-roam-backlinks-get (org-roam-node-at-point) :unique t)))
      (save-excursion
        (when heading-func
          (funcall heading-func))
        (insert " 反向链接\n")
        (dolist (backlink backlinks)
          (let* ((source-node (org-roam-backlink-source-node backlink))
                 (point (org-roam-backlink-point backlink)))
            (insert
             (format "- [[id:%s][%s]]\n"
                     (org-roam-node-id source-node)
                     (org-roam-node-title source-node))))))))

(defun my/insert-backlinks ()
  (goto-char (point-min))
  (org-shifttab 0)
  (while (= 0 (org-next-visible-heading 1))
    (when (assoc "ID" (org-entry-properties))
      (my/insert--backlinks
       (lambda ()
         (org-insert-subheading t)))))
  (goto-char (point-min))
  (my/insert--backlinks
   (lambda ()
     (goto-char (point-max))
     (insert "* "))))

(defun my/extract-org-roam-id (origin-filename)
  (save-excursion
    (goto-char (point-min))
    (let ((matche '()))
      (while (re-search-forward "\\[\\[id:\\(.*?\\)]\\[.*?\\]\\]" nil t)
        (when-let* ((uuid (match-string-no-properties 1))
                    (roam-node (org-roam-node-from-id uuid))
                    (file-name (car (org-roam-id-find uuid))))
          (when (member "publish" (org-roam-node-tags roam-node))
            (push file-name matche))))
      (delete origin-filename (seq-uniq matche)))))

(defun my/org-roam-publish (&optional skip-publish skip-update-backlinks)
  "Publish current file"
  (interactive)
  (goto-char (point-min))
  (org-roam-update-org-id-locations)
  (org-roam-tag-add (list "publish"))
  (unless skip-update-backlinks

config.el
(setq doom-theme 'ef-spring)

(setq user-full-name "SouthFox"
      user-mail-address "master@southfox.me")

(defun doom-dashboard-draw-ascii-emacs-banner-fn ()
  (let* ((banner
          '("                                                                   ,           "
            "                                                             _.-=;~ /_         "
            "                                                          _-~   '     ;.       "
            "                                                      _.-~     '   .-~-~`-._   "
            "                                                _.--~~:.             --.____88 "
            "                              ____.........--~~~. .' .  .        _..-------~~  "
            "                     _..--~~~~               .' .'             ,'              "
            "                 _.-~                        .       .     ` ,'                "
            "               .'                                    :.    ./                  "
            "             .:     ,/          `                   ::.   ,'                   "
            "           .:'     ,(            ;.                ::. ,-'                     "
            "          .'     ./'.`.     . . /:::._______.... _/:.o/                        "
            "         /     ./'. . .)  . _.,'               `88;?88|                        "
            "       ,'  . .,/'._,-~ /_.o8P'                  88P ?8b                        "
            "    _,'' . .,/',-~    d888P'                    88'  88|                       "
            " _.'~  . .,:oP'        ?88b              _..--- 88.--'8b.--..__                "
            ":     ...' 88o __,------.88o ...__..._.=~- .    `~~   `~~      ~-._ Fox! _.    "
            "`.;;;:='    ~~            ~~~                ~-    -       -   -               "))
         (longest-line (apply #'max (mapcar #'length banner))))
    (put-text-property
     (point)
     (dolist (line banner (point))
       (insert (+doom-dashboard--center
                +doom-dashboard--width
                (concat
                 line (make-string (max 0 (- longest-line (length line)))
                                   32)))
               "\n"))
     'face 'doom-dashboard-banner)))

;(unless (display-graphic-p) ; for some reason this messes up the graphical splash screen atm
  (setq +doom-dashboard-ascii-banner-fn #'doom-dashboard-draw-ascii-emacs-banner-fn) ;)

(after! org
  (setq org-directory "~/org/"
        org-image-actual-width '(400)
        org-link-search-must-match-exact-headline nil
        org-log-done 'time
        org-log-into-drawer t))

(use-package! org-modern
  :hook (org-mode . org-modern-mode)
  :config
  (setq org-modern-star '("◉" "○" "✸" "✿" "✤" "✜" "◆" "▶")
        org-modern-table-vertical 1
        org-modern-table-horizontal 0.2
        org-modern-list '((43 . "➤")
                          (45 . "–")
                          (42 . "•"))
        org-modern-block-name
        '((t . t)
          ("src" "»" "«")
          ("example" "»–" "–«")
          ("quote" "❝" "❞")
          ("export" "⏩" "⏪"))
        org-modern-progress nil
        org-modern-priority nil
        org-modern-horizontal-rule (make-string 36 ?─)
        )
  )

(setq org-agenda-files '("~/Sync/org/GTD/inbox.org"
                         "~/Sync/org/GTD/gtd.org"
                         "~/Sync/org/GTD/tickler.org"))

(after! org
  (setq org-capture-templates '(("t" "Todo [inbox]" entry
                               (file "~/Sync/org/GTD/inbox.org")
                               "* TODO %i%? \n SCHEDULED: %t")
                              ("T" "Tickler" entry
                               (file+headline "~/Sync/org/GTD/tickler.org" "Tickler")
                               "* %i%? \n %U"))))

(setq org-roam-directory "~/Sync/org/Note/org-roam")
(setq org-roam-capture-templates
      '(("m" "main" plain
	 "%?"
	 :if-new (file+head "main/%<%Y%m%d%H%M%S>-${slug}.org"
			    "#+title: ${title}\n#+date: %T\n#+hugo_auto_set_lastmod: t\n#+hugo_section: main\n")
	 :immediate-finish t
	 :unnarrowed t)
	("r" "references" plain "%?"
	 :if-new
	 (file+head "references/%<%Y%m%d%H%M%S>-${title}.org" "#+title: ${title}\n#+date: %T\n#+hugo_auto_set_lastmod: t\n#+hugo_section: references\n")
	 :immediate-finish t
	 :unnarrowed t)
	("a" "article" plain "%?"
	 :if-new
	 (file+head "articles/%<%Y%m%d%H%M%S>-${title}.org" "#+title: ${title}\n#+date: %T\n#+filetags: :article:\n#+hugo_auto_set_lastmod: t\n#+hugo_section: articles\n")
	 :immediate-finish t
	 :unnarrowed t)))

(setq org-roam-dailies-capture-templates
      '(("d" "default" entry
         "* %?"
         :target (file+head "%<%Y-%m-%d>.org"
                            "#+title: %<%Y-%m-%d>\n#+date: %T\n#+hugo_auto_set_lastmod: t\n#+hugo_section: daily\n"))))

(use-package! websocket
    :after org-roam)

(use-package! org-roam-ui
    :after org-roam ;; or :after org
;         normally we'd recommend hooking orui after org-roam, but since org-roam does not have
;         a hookable mode anymore, you're advised to pick something yourself
;         if you don't care about startup time, use
;  :hook (after-init . org-roam-ui-mode)
    :config
    (setq org-roam-ui-sync-theme t
          org-roam-ui-follow t
          org-roam-ui-update-on-save t
          org-roam-ui-open-on-start t))

(use-package! org-fragtog
  :hook (org-mode . org-fragtog-mode))

(setq org-hugo-base-dir "~/Documents/roam-publish/")

(defun my/org-roam-filter-by-tag (tag-name)
  (lambda (node)
    (member tag-name (org-roam-node-tags node))))

(defun my/org-roam-list-notes-by-tag (tag-name)
  (mapcar #'org-roam-node-file
          (seq-filter
           (daviwil/org-roam-filter-by-tag tag-name)
           (org-roam-node-list))))

(defun my/org-roam-export-all ()
  "Re-exports all Org-roam files to Hugo markdown."
  (interactive)
  (dolist (org-file (my/org-roam-list-notes-by-tag "publish"))
  ;(dolist (org-file (directory-files-recursively org-roam-directory "\.org$"))
    (with-current-buffer (find-file org-file)
        (org-hugo-export-wim-to-md))))

(defun my/org-roam-publish ()
  "Publish current file"
  (interactive)
  (org-roam-update-org-id-locations)
  (org-roam-set-keyword "filetags" ":publish:")
  (save-buffer)
  (org-hugo-export-wim-to-md)
  (async-shell-command (concat
                        "cd " org-hugo-base-dir
                        " && " "git add ."
                        (if (yes-or-no-p "Push now?")
                            (concat " && " "git commit -m '[post] new post'"
                                    " && " "git push"))) "*Messages*")
  (message "publish nwe post!"))

(defun my/org-roam-creat-node ()
  "creat node and add IS_NODE property"
  (interactive)
  (org-id-get-create)
  (org-set-tags ":NODE:")
  (save-buffer)
  (org-hugo-export-wim-to-md))

(org-babel-do-load-languages
 'org-babel-load-languages
 '((restclient . t)))

(setq cnfonts-personal-fontnames
      '(()
        ()
        ("Noto Color Emoji")
        ()
        ()))
(cnfonts-mode 1)

(setq word-wrap-by-category t)

(setq doom-unicode-font (font-spec :family "Noto Color Emoji"))
(add-to-list 'default-frame-alist '(height . 35))
(add-to-list 'default-frame-alist '(width . 102))

(add-hook 'python-mode-hook  #'rainbow-delimiters-mode)
(add-hook 'clojure-mode-hook #'rainbow-delimiters-mode)

(setq shell-file-name "/bin/zsh"
      vterm-max-scrollback 5000)

(parrot-mode 1)
(define-key evil-normal-state-map (kbd "[r") 'parrot-rotate-prev-word-at-point)
(define-key evil-normal-state-map (kbd "]r") 'parrot-rotate-next-word-at-point)

(setq parrot-rotate-dict
      '(
        (:rot ("yes" "no") :caps t :upcase t)
        (:rot ("&" "|"))
        (:rot ("begin" "end") :caps t :upcase t)
        (:rot ("enable" "disable") :caps t :upcase t)
        (:rot ("enter" "exit") :caps t :upcase t)
        (:rot ("forward" "backward") :caps t :upcase t)
        (:rot ("front" "rear" "back") :caps t :upcase t)
        (:rot ("get" "set") :caps t :upcase t)
        (:rot ("high" "low") :caps t :upcase t)
        (:rot ("in" "out") :caps t :upcase t)
        (:rot ("left" "right") :caps t :upcase t)
        (:rot ("min" "max") :caps t :upcase t)
        (:rot ("on" "off") :caps t :upcase t)
        (:rot ("start" "stop") :caps t :upcase t)
        (:rot ("true" "false") :caps t :upcase t)
        (:rot ("&&" "||"))
        (:rot ("==" "!="))
        (:rot ("if" "else" "elif"))
        (:rot ("ifdef" "ifndef"))
        ))

(defun my-add-to-multiple-hooks (function hooks)
  (mapc (lambda (hook)
          (add-hook hook function))
        hooks))

(add-hook 'emacs-startup-hook
          (lambda ()
          (setq parrot-num-rotations 10)
          (parrot-set-parrot-type 'emacs)))

(defun my-parrot-thumbsup-play ()
  (parrot-set-parrot-type 'thumbsup)
  (parrot-start-animation))

(defun my-parrot-emacs-play ()
  (parrot-set-parrot-type 'emacs)
  (parrot-start-animation))

(defun my-parrot-rotating-play ()
  (parrot-set-parrot-type 'rotating)
  (parrot-start-animation))

(my-add-to-multiple-hooks
 'my-parrot-thumbsup-play
 '(org-after-todo-state-change-hook
   org-clock-in-hook
   org-timer-done-hook
   git-commit-post-finish-hook))

(my-add-to-multiple-hooks
 'my-parrot-emacs-play
 '(after-save-hook
   find-file-hook))

(my-add-to-multiple-hooks
 'my-parrot-rotating-play
 '(elfeed-search-mode-hook
   mu4e-main-mode-hook))

(load! "secrets")
(load! "elisp/hy")
(setq-default custom-file (expand-file-name "secrets.el" doom-user-dir))
(when (file-exists-p custom-file)
  (load custom-file))

(setq bibtex-completion-pdf-field "File")
(setq bibtex-completion-bibliography '("~/Sync/Ebook/catalog.bib"))
(setq citar-bibliography '("~/Sync/Ebook/catalog.bib"))

(setq calibredb-root-dir "~/Documents/Ebook")
(setq calibredb-db-dir (expand-file-name "metadata.db" calibredb-root-dir))
(setq calibredb-ref-default-bibliography (concat (file-name-as-directory calibredb-root-dir) "catalog.bib"))
(setq calibredb-format-all-the-icons t)

(map! :map calibredb-show-mode-map
      :n "?" #'calibredb-entry-dispatch
      :n "o" #'calibredb-find-file
      :n "O" #'calibredb-find-file-other-frame
      :n "V" #'calibredb-open-file-with-default-tool
      :n "s" #'calibredb-set-metadata-dispatch
      :n "e" #'calibredb-export-dispatch
      :n "q" #'calibredb-entry-quit
      :n "." #'calibredb-open-dired
      :n [tab] #'calibredb-toggle-view-at-point
      :n "M-t" #'calibredb-set-metadata--tags
      :n "M-a" #'calibredb-set-metadata--author_sort
      :n "M-A" #'calibredb-set-metadata--authors
      :n "M-T" #'calibredb-set-metadata--title
      :n "M-c" #'calibredb-set-metadata--comments)
(map! :map calibredb-search-mode-map
      :n [mouse-3] #'calibredb-search-mouse
      :n "RET" #'calibredb-find-file
      :n "?" #'calibredb-dispatch
      :n "a" #'calibredb-add
      :n "A" #'calibredb-add-dir
      :n "c" #'calibredb-clone
      :n "d" #'calibredb-remove
      :n "D" #'calibredb-remove-marked-items
      :n "j" #'calibredb-next-entry
      :n "k" #'calibredb-previous-entry
      :n "l" #'calibredb-virtual-library-list
      :n "L" #'calibredb-library-list
      :n "n" #'calibredb-virtual-library-next
      :n "N" #'calibredb-library-next
      :n "p" #'calibredb-virtual-library-previous
      :n "P" #'calibredb-library-previous
      :n "s" #'calibredb-set-metadata-dispatch
      :n "S" #'calibredb-switch-library
      :n "o" #'calibredb-find-file
      :n "O" #'calibredb-find-file-other-frame
      :n "v" #'calibredb-view
      :n "V" #'calibredb-open-file-with-default-tool
      :n "." #'calibredb-open-dired
      :n "b" #'calibredb-catalog-bib-dispatch
      :n "e" #'calibredb-export-dispatch
      :n "r" #'calibredb-search-refresh-and-clear-filter
      :n "R" #'calibredb-search-clear-filter
      :n "q" #'calibredb-search-quit
      :n "m" #'calibredb-mark-and-forward
      :n "f" #'calibredb-toggle-favorite-at-point
      :n "x" #'calibredb-toggle-archive-at-point
      :n "h" #'calibredb-toggle-highlight-at-point
      :n "u" #'calibredb-unmark-and-forward
      :n "i" #'calibredb-edit-annotation
      :n "DEL" #'calibredb-unmark-and-backward
      :n [backtab] #'calibredb-toggle-view
      :n [tab] #'calibredb-toggle-view-at-point
      :n "M-n" #'calibredb-show-next-entry
      :n "M-p" #'calibredb-show-previous-entry
      :n "/" #'calibredb-search-live-filter
      :n "M-t" #'calibredb-set-metadata--tags
      :n "M-a" #'calibredb-set-metadata--author_sort
      :n "M-A" #'calibredb-set-metadata--authors
      :n "M-T" #'calibredb-set-metadata--title
      :n "M-c" #'calibredb-set-metadata--comments)

(setq sdcv-say-word-p t)
(setq   sdcv-dictionary-data-dir (expand-file-name "~/.stardict/dic/"))

(map! :leader :desc "sdvc" "z" #'sdcv-search-pointer+)

(map! :leader
      (:prefix ("c h" . "Help info from Clippy")
       :desc "Clippy describes function under point" "f" #'clippy-describe-function
       :desc "Clippy describes variable under point" "v" #'clippy-describe-variable))

(map! :leader
      :prefix ("o")
      :desc "Mastodon"          "M" #'mastodon)

(map! :after mastodon
      :map mastodon-mode-map
      :n "[ [" #'mastodon-tl--goto-prev-toot
      :n "] ]" #'mastodon-tl--goto-next-toot
      :n "g k" #'mastodon-tl--previous-tab-item
      :n "g j" #'mastodon-tl--next-tab-item

      :n "q" #'kill-current-buffer
      :n "Q" #'kill-buffer-and-window

      ;;; timelines
      :n "#" #'mastodon-tl--get-tag-timeline
      :n "A" #'mastodon-profile--get-toot-author
      :n "F" #'mastodon-tl--get-federated-timeline
      :n "H" #'mastodon-tl--get-home-timeline
      :n "L" #'mastodon-tl--get-local-timeline
      :n "N" #'mastodon-notifications-get
      :n "O" #'mastodon-profile--my-profile
      :n "P" #'mastodon-profile--show-user
      :n "T" #'mastodon-tl--thread

      ;;; toot actions
      :n "K" #'mastodon-toot--bookmark-toot-toggle
      :n "R" #'mastodon-toot--toggle-boost
      :n "c" #'mastodon-tl--toggle-spoiler-text-in-toot
      :n "C" #'mastodon-toot--copy-toot-url
      :n "o" #'mastodon-url-lookup
      :n "d" #'mastodon-toot--delete-toot
      :n "D" #'mastodon-toot--delete-draft-toot
      :n "f" #'mastodon-toot--toggle-favourite
      :n "r" #'mastodon-toot--reply
      :n "u" #'mastodon-tl--update
      :n "v" #'mastodon-tl--poll-vote

      ;;; toot!
      :n "t" #'mastodon-toot

      ;;; mastodon additions
      :n "S"    #'mastodon-search--search-query
      :n "V F"  #'mastodon-profile--view-favourites
      :n "V B"  #'mastodon-profile--view-bookmarks
      :n "V L" #'mastodon-tl--view-list-timeline
      )

(beacon-mode 1)

(setq ement-save-sessions t
      ement-room-message-format-spec "%B%r%R%t")
(map! :map ement-room-mode-map
      :n "c e" #'ement-room-edit-message
      :n "c m" #'ement-room-send-message
      :n "c r" #'ement-room-send-reaction
      :n "c i" #'ement-room-send-image
      :n "c f" #'ement-room-send-file
      :n "q"   #'kill-current-buffer
      )
(map! :map ement-room-list-mode-map
      :n "RET" #'ement-room-list-RET
      :n "q"   #'kill-current-buffer
      )

(after! dap-mode
  (setq dap-python-debugger 'debugpy))

(map! :map dap-mode-map
      :leader
      :prefix ("d" . "dap")
      ;;; basics
      :desc "dap next"          "n" #'dap-next
      :desc "dap step in"       "i" #'dap-step-in
      :desc "dap step out"      "o" #'dap-step-out
      :desc "dap continue"      "c" #'dap-continue
      :desc "dap hydra"         "h" #'dap-hydra
      :desc "dap debug restart" "r" #'dap-debug-restart
      :desc "dap debug"         "s" #'dap-debug

      ;;; debug
      :prefix ("dd" . "Debug")
      :desc "dap debug recent"  "r" #'dap-debug-recent
      :desc "dap debug last"    "l" #'dap-debug-last

      ;;; eval
      :prefix ("de" . "Eval")
      :desc "eval"                "e" #'dap-eval
      :desc "eval region"         "r" #'dap-eval-region
      :desc "eval thing at point" "s" #'dap-eval-thing-at-point
      :desc "add expression"      "a" #'dap-ui-expressions-add
      :desc "remove expression"   "d" #'dap-ui-expressions-remove

      :prefix ("db" . "Breakpoint")
      :desc "dap breakpoint toggle"      "b" #'dap-breakpoint-toggle
      :desc "dap breakpoint condition"   "c" #'dap-breakpoint-condition
      :desc "dap breakpoint hit count"   "h" #'dap-breakpoint-hit-condition
      :desc "dap breakpoint log message" "l" #'dap-breakpoint-log-message)

(use-package! rime
  :custom
  (default-input-method "rime")
  :config
  (define-key rime-mode-map (kbd "C-i") 'rime-force-enable)
  (setq rime-show-candidate 'posframe)
  (setq rime-disable-predicates
        '(rime-predicate-evil-mode-p
          rime-predicate-after-alphabet-char-p
          rime-predicate-space-after-cc-p
          rime-predicate-current-uppercase-letter-p
          rime-predicate-punctuation-line-begin-p)))

(setq leetcode-save-solutions t)
(setq leetcode-directory "~/Documents/leetcode")

(after! company
  (setq company-idle-delay 0.5
        company-minimum-prefix-length 2)
  (setq company-show-numbers t))

(use-package! screenshot
  :defer t)

(set-email-account! "southfox.me"
  '((mu4e-sent-folder       . "/Sent")
    (mu4e-drafts-folder     . "/southfox.me/Drafts")
    (mu4e-trash-folder      . "/southfox.me/Trash")
    (mu4e-refile-folder     . "/southfox.me/All Mail")
    (smtpmail-smtp-user     . "master@southfox.me")
    (mu4e-compose-signature . "---\nFor mu4e"))
  t)

(add-hook! 'elfeed-search-mode-hook 'elfeed-update)

(after! elfeed
  (setq elfeed-search-filter "@6-month-ago +unread")
  (setq rmh-elfeed-org-files '("~/Sync/org/rss.org"))
  (setq elfeed-curl-extra-arguments '("-H Mozilla/5.0 (Windows NT 10.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.7113.93 Safari/537.36"
                                      "--proxy" "socks5://127.0.0.1:10808"
                                      "--retry" "2"
                                      "--insecure")))

(use-package! flycheck-clj-kondo
  :after (clojure-mode))

(use-package! clj-refactor
  :after (clojure-mode))

(use-package! paredit
  :after (clojure-mode)
  :config
  (add-hook 'clojure-mode-hook 'enable-paredit-mode)
  (add-hook 'emacs-lisp-mode-hook 'enable-paredit-mode)
  (add-hook 'cider-repl-mode-hook 'enable-paredit-mode))
